File: AdminUi/packages/admin_api_sdk/lib/src/endpoints/clients_endpoint.dart
import 'package:admin_api_types/admin_api_types.dart';

import '../types/types.dart';
import 'endpoint.dart';

class ClientsEndpoint extends Endpoint {
  ClientsEndpoint(super.dio);

  Future<ApiResponse<List<Clients>>> getClients() => get(
        '/api/v1/Clients',
        transformer: (e) => (e as List).map(Clients.fromJson).toList(),
      );

  Future<ApiResponse<Client>> createClient({
    required String defaultTier,
    String? clientId,
    String? clientSecret,
    String? displayName,
    int? maxIdentities,
  }) =>
      post(
        '/api/v1/Clients',
        data: {
          'defaultTier': defaultTier,
          'clientId': clientId,
          'clientSecret': clientSecret,
          'displayName': displayName,
          'maxIdentities': maxIdentities,
        },
        transformer: Client.fromJson,
      );

  Future<ApiResponse<Client>> getClient(String clientId) => get(
        '/api/v1/Clients/$clientId',
        transformer: Client.fromJson,
      );

  Future<ApiResponse<Client>> changeClientSecret(String clientId, {required String? newSecret}) => patch(
        '/api/v1/Clients/$clientId/ChangeSecret',
        data: {
          'newSecret': newSecret,
        },
        transformer: Client.fromJson,
      );

  Future<ApiResponse<Client>> updateClient(
    String clientId, {
    required String defaultTier,
    required int? maxIdentities,
  }) =>
      put(
        '/api/v1/Clients/$clientId',
        data: {
          'defaultTier': defaultTier,
          'maxIdentities': maxIdentities,
        },
        transformer: Client.fromJson,
      );

  Future<ApiResponse<void>> deleteClient(String clientId) => delete(
        '/api/v1/Clients/$clientId',
        expectedStatus: 204,
        transformer: (e) {},
        allowEmptyResponse: true,
      );
}

File: AdminApi.Sdk/Endpoints/Clients/ClientsEndpoint.cs
using Backbone.AdminApi.Sdk.Endpoints.Clients.Types;
using Backbone.AdminApi.Sdk.Endpoints.Clients.Types.Requests;
using Backbone.AdminApi.Sdk.Endpoints.Clients.Types.Responses;
using Backbone.BuildingBlocks.SDK.Endpoints.Common;
using Backbone.BuildingBlocks.SDK.Endpoints.Common.Types;

namespace Backbone.AdminApi.Sdk.Endpoints.Clients;

public class ClientsEndpoint(EndpointClient client) : AdminApiEndpoint(client)
{
    public async Task<ApiResponse<ListClientsResponse>> GetAllClients() => await _client.Get<ListClientsResponse>($"api/{API_VERSION}/Clients");

    public async Task<ApiResponse<ClientInfo>> GetClient(string id) => await _client.Get<ClientInfo>($"api/{API_VERSION}/Clients/{id}");

    public async Task<ApiResponse<CreateClientResponse>> CreateClient(CreateClientRequest request)
        => await _client.Post<CreateClientResponse>($"api/{API_VERSION}/Clients", request);

    public async Task<ApiResponse<ClientInfo>> ChangeClientSecret(string id, ChangeClientSecretRequest request)
        => await _client.Patch<ClientInfo>($"api/{API_VERSION}/Clients/{id}/ChangeSecret", request);

    public async Task<ApiResponse<ClientInfo>> UpdateClient(string id, UpdateClientRequest request)
        => await _client.Put<ClientInfo>($"api/{API_VERSION}/Clients/{id}", request);

    public async Task<ApiResponse<EmptyResponse>> DeleteClient(string id) => await _client.Delete<EmptyResponse>($"api/{API_VERSION}/Clients/{id}");
}
